Source/Task/WaitTimer_stl.cpp

```diff
@@ -0,0 +1,136 @@
+#include "pch.h"
+#include "WaitTimer.h"
+
+using Deadline = std::chrono::high_resolution_clock::time_point;
+
+class WaitTimerImpl
+{
+public:
+
+    ~WaitTimerImpl()
+    {
+        if (m_thread.joinable())
+        {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            m_abortThread = true;
+            m_cv.notify_all();
+        }
+
+        m_thread.join();
+    }
+
+    HRESULT Initialize(_In_opt_ void* context, _In_ WaitTimerCallback* callback)
+    {
+        m_context = context;
+        m_callback = callback;
+
+        try
+        {
+            std::thread t([this] { Worker(); });
+            m_thread.swap(t);
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+
+        return S_OK;
+    }
+
+    void Start(_In_ uint64_t absoluteTime)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_duration = Deadline::duration(absoluteTime);
+        m_deadlineSet = true;
+        m_cv.notify_all();
+    }
+
+    void Cancel()
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_deadlineSet = false;
+        m_cv.notify_all();
+    }
+
+private:
+
+    void* m_context;
+    WaitTimerCallback* m_callback;
+    Deadline::duration m_duration;
```

Comment:
why not store it as a time point (Deadline), this value is set to an absolute point in time (and later you have to do a whole bunch of casting)

Reply:
Yup, cleaner.  Updated.